plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,284 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class LdapGroupProvider
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupProvider.class);
+    private final LdapClient client;
+    protected final List<String> userBindSearchPatterns;
+    protected final Optional<String> userBaseDistinguishedName;
+    protected final Optional<String> groupBaseDistinguishedName;
+    protected final Optional<String> groupNameAttribute;
+    protected final Optional<String> groupMembershipAttribute;
+    protected final Optional<String> groupUserMembershipAttribute;
+    protected final String bindDistinguishedName;
+    protected final String bindPassword;
+    protected final boolean allowUserNotExist;
+    private final LoadingCache<String, Set<String>> groupCache;
+    private Optional<LdapGroupProviderQuery> ldapGroupProviderQueryUser;
+    private Optional<LdapGroupProviderQuery> ldapGroupProviderQueryGroup;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapGroupProviderConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBindSearchPatterns = ldapConfig.getUserBindSearchPatterns();
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = Optional.ofNullable(ldapConfig.getGroupBaseDistinguishedName());
+        this.groupNameAttribute = Optional.ofNullable(ldapConfig.getGroupNameAttribute());
+        this.bindDistinguishedName = ldapConfig.getBindDistinguishedName();
+        this.bindPassword = ldapConfig.getBindPassword();
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+        this.groupMembershipAttribute = Optional.ofNullable(ldapConfig.getGroupMembershipAttribute());
+        this.groupUserMembershipAttribute = Optional.ofNullable(ldapConfig.getGroupUserMembershipAttribute());
+        if (this.groupMembershipAttribute.isEmpty() && this.groupUserMembershipAttribute.isEmpty()) {
+            throw new LdapGroupMembershipException("both ldap.group-membership-attribute and "
+                    + "ldap.group-user-membership-attribute cannot be empty");
+        }
+        if (this.groupMembershipAttribute.isPresent() && this.groupNameAttribute.isEmpty()) {
+            throw new LdapGroupMembershipException("ldap.group-membership-attribute is set "
+                    + "ldap.group-name-attribute cannot be empty");
+        }
+        this.groupCache = buildNonEvictableCacheWithWeakInvalidateAll(
+                CacheBuilder.newBuilder()
+                        .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS),
+                CacheLoader.from(this::getGroupsWithBindDistinguishedName));
```

Comment:
Note: I'm intending to add a generic caching layer in Trino for group providers. See #18174. I'll take some inspiration from how you did this here :) If that caching layer is accepted, caching here would be unnecessary.